marker/processors/high_quality_text.py

```diff
@@ -0,0 +1,207 @@
+You will receive an image of a text block and a set of extracted lines corresponding to the text in the image.
+Your task is to correct any errors in the extracted lines, including math, formatting, and other inaccuracies, and output the corrected lines in a JSON format.
+The number of output lines MUST match the number of input lines.
+
+**Instructions:**
+
+1. Carefully examine the provided text block image .
+2. Analyze the extracted lines.
+3. For each extracted line, compare it to the corresponding line in the image.
+4. Correct any errors in the extracted line, including:
+    * Inline math: Ensure all mathematical expressions are correctly formatted and rendered.
+    * Formatting: Maintain consistent formatting with the text block image, including spacing, indentation, and special characters.
+    * Other inaccuracies:  If the image is handwritten then you may correct any spelling errors, or other discrepancies.
+5. Do not remove any formatting i.e bold, italics, etc from the extracted lines unless it is necessary to correct the error.
+6. Ensure that inline math is properly enclosed in dollar signs.
+7. The number of corrected lines in the output MUST equal the number of extracted lines provided in the input. Do not add or remove lines.
+8. Output the corrected lines in JSON format with a "lines" field, as shown in the example below.
+
+**Example:**
+
+Input:
+```
+{
+ "extracted_lines": [
+  "Adversarial training (AT) [23], which aims to minimize\n",
+  "the model's risk under the worst-case perturbations, is cur-\n",
+  "rently the most effective approach for improving the robust-\n",
+  "ness of deep neural networks. For a given neural network\n",
+  "f(x, w) with parameters w, the optimization objective of\n",
+  "AT can be formulated as follows:\n"
+ ]
+}
+```
+
+Output:
+
+```json
+{
+ "corrected_lines": [
+  "Adversarial training (AT) [23], which aims to minimize\n",
+  "the model's risk under the worst-case perturbations, is cur-\n",
+  "rently the most effective approach for improving the robust-\n",
+  "ness of deep neural networks. For a given neural network\n",
+  "$f(x, w)$ with parameters $w$, the optimization objective of\n",
+  "AT can be formulated as follows:\n"
+ ]
+}
+```
+
+**Input:**
+
+"""
+
+    def __init__(self, config=None):
+        super().__init__(config)
+
+        self.model = None
+        if not self.high_quality:
+            return
+
+        if self.google_api_key is None:
+            raise ValueError("Google API key is not set")
+
+        genai.configure(api_key=self.google_api_key)
+        self.model = genai.GenerativeModel(self.model_name)
+
+    def __call__(self, document: Document):
+        if not self.high_quality or self.model is None:
+            return
+
+        self.rewrite_blocks(document)
+
+    def rewrite_blocks(self, document: Document):
+        pbar = tqdm(desc="High quality text processor")
+        with ThreadPoolExecutor() as executor:
```

Comment:
We should probably limit the number of workers so we don't get too much concurrency.  Or at least make it configurable.

Reply:
makes sense!